frodo.js

```diff
@@ -1,66 +1,61 @@
 /**
```

Comment:
SCRIPT-I'm impressed with your use of design patterns here, it really enhances the code organization.